File: python/longest_substring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        counter = {}
        ans = 0
        i = 0 # sliding window

        for j in range(len(s)):
            chr = s[j]

            if (chr in counter):
                i = max(counter[chr], i)

            ans = max(ans, j + 1 - i)
            counter[chr] = j + 1
        
        return ans

# test here
print(Solution().lengthOfLongestSubstring('dvdf'))

File: javascript/longest_substring.js
var lengthOfLongestSubstring = function(s) {
    let chars = {}
    let j = 0; // start of substr
    let len = 0;

    for (i = 0; i < s.length; i++) {
        let char = chars?.[s[i]];

        if (char === undefined) {
            chars[s[i]] = 0;
        } // if

        j = Math.max(chars[s[i]], j);
        len = Math.max(len, i - j + 1);
        chars[s[i]] = i + 1;
    } // for

    return len;
};

// test here
console.log(lengthOfLongestSubstring('pwwkew'));
